presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalCteOptimizer.java

```diff
@@ -0,0 +1,301 @@
+            for (VariableReferenceExpression oldVariable : tempScan.getOutputVariables()) {
+                VariableReferenceExpression newVariable = variableAllocator.newVariable(oldVariable);
+                newOutputVariables.add(newVariable);
+                newColumnAssignmentsMap.put(newVariable, tempScan.getAssignments().get(oldVariable));
+            }
+
+            TableScanNode tableScanNode = new TableScanNode(
+                    Optional.empty(),
+                    idAllocator.getNextId(),
+                    tempScan.getTable(),
+                    newOutputVariables,
+                    newColumnAssignmentsMap,
+                    tempScan.getCurrentConstraint(),
+                    tempScan.getEnforcedConstraint());
+
+            /* The temporary table scan might have columns removed by the UnaliasSymbolReferences and other optimizers (its a plan tree after all),
+            use originalOutputVariables (which are also canonicalized and maintained) and add them back */
+            Map<VariableReferenceExpression, VariableReferenceExpression> intermediateReferenceMap = new HashMap<>();
+            for (int i = 0; i < tempScan.getOutputVariables().size(); i++) {
+                intermediateReferenceMap.put(tempScan.getOutputVariables().get(i), newOutputVariables.get(i));
+            }
+
+            Assignments.Builder assignments = Assignments.builder();
+            for (int i = 0; i < tableInfo.getOriginalOutputVariables().size(); i++) {
+                assignments.put(node.getOutputVariables().get(i), intermediateReferenceMap.get(tableInfo.getOriginalOutputVariables().get(i)));
+            }
+            return new ProjectNode(Optional.empty(), idAllocator.getNextId(), Optional.empty(),
+                    tableScanNode, assignments.build(), ProjectNode.Locality.LOCAL);
+        }
+
+        public boolean isPlanRewritten()
+        {
+            return isPlanRewritten;
+        }
+    }
+
+    public static class PhysicalCteTransformerContext
+    {
+        private Map<String, TemporaryTableInfo> cteNameToTableInfo = new HashMap<>();
+
+        public PhysicalCteTransformerContext()
+        {
+            cteNameToTableInfo = new HashMap<>();
+        }
+
+        public void put(String cteName, TemporaryTableInfo handle)
+        {
+            cteNameToTableInfo.put(cteName, handle);
+        }
+
+        public TemporaryTableInfo getTableInfo(String cteName)
+        {
+            return cteNameToTableInfo.get(cteName);
+        }
+
+        public static class TemporaryTableInfo
+        {
+            private final TableScanNode tableScanNode;
+            private final List<VariableReferenceExpression> originalOutputVariables;
+
+            public TemporaryTableInfo(TableScanNode tableScanNode, List<VariableReferenceExpression> originalOutputVariables)
+            {
+                this.tableScanNode = tableScanNode;
+                this.originalOutputVariables = originalOutputVariables;
```

Comment:
Please check inputs are not null, and create an immutable copy of the original output variables (no cost if the input is immutable as well)